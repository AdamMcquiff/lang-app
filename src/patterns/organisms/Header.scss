.o-header {
  padding: 1em;
}