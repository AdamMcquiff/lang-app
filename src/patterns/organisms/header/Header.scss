.o-header {
  padding: 1em;

  h2 {
    @include type-setting(1);
  }
}