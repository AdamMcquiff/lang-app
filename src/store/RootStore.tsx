import SettingsService from '../services/settingsService/SettingsService';
import TranslationCategoryService from '../services/translationCategoriesService/TranslationCategoryService';
import TranslationService from '../services/translationService/TranslationService';
import SettingsStore from './SettingsStore';
import TranslationCategoryStore from './TranslationCategoryStore';
import TranslationStore from './TranslationStore';

class RootStore {
	public settingsStore: SettingsStore;
	public translationCategoryStore: TranslationCategoryStore;
	public translationStore: TranslationStore;

	constructor() {
		this.settingsStore = new SettingsStore(this, SettingsService);
		this.translationCategoryStore = new TranslationCategoryStore(this, TranslationCategoryService);
		this.translationStore = new TranslationStore(this, TranslationService);
	}
}

export default RootStore;