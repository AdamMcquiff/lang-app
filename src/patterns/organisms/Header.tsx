import * as React from 'react';
import RootStore from '../../store/RootStore';
import TextForm from '../molecules/TextForm';
import './Header.scss';

interface IHeaderProps {
    store: RootStore
}

interface IHeaderState {
    addCategoryTextValue: string
}

class Header extends React.Component<IHeaderProps, IHeaderState> {
    public state: IHeaderState = {
        addCategoryTextValue: ""
    }

    constructor(props: IHeaderProps) {
        super(props);
        this.onAddCategoryButtonClick = this.onAddCategoryButtonClick.bind(this);
        this.onAddCategoryTextChange = this.onAddCategoryTextChange.bind(this);
    }

    public onAddCategoryButtonClick() {
        this.props.store.translationStore.createTranslationCategory({ 
            name: this.state.addCategoryTextValue 
        });
    }
  
    public onAddCategoryTextChange(event: any) {
        this.setState({ addCategoryTextValue: event.target.value })
    }

    public render() {
        return (
            <header className="o-header">
                <TextForm buttonA11yLabel="Add new category"
                          buttonTextLabel="Add"
                          inputLabel="New Category" 
                          onButtonClick={this.onAddCategoryButtonClick}
                          onTextChange={this.onAddCategoryTextChange}/>
            </header>
        )
    }
}

export default Header;