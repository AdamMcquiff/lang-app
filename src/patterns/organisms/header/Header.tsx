import { observer } from 'mobx-react';
import * as React from 'react';
import ITranslationCategoryModel from '../../../models/TranslationCategory';
import RootStore from '../../../store/RootStore';
import TextForm from '../../molecules/textForm/TextForm';
import './Header.scss';

interface IHeaderProps {
    store: RootStore,
    className?: string
}

interface IHeaderState {
    addCategoryTextValue: string
}

@observer
class Header extends React.Component<IHeaderProps, IHeaderState> {
    public state: IHeaderState = {
        addCategoryTextValue: ''
    }

    public onAddCategorySubmit = () => {
        this.props.store.translationCategoryStore.create({ 
            title: this.state.addCategoryTextValue 
        });
    }
  
    public onAddCategoryTextChange = (event: any) => {
        this.setState({ addCategoryTextValue: event.target.value })
    }

    public render() {
        return (
            <header className={'o-header ' + this.props.className}>
                <h2>Categories</h2>
                
                <ul>
                    {this.props.store.translationCategoryStore.translationCategories.map((category: ITranslationCategoryModel, id: number) => {
                        return <li key={id}>
                            {category.title}
                        </li>
                    })}
                </ul>

                <TextForm hasButton={false}
                          inputLabel="New Category" 
                          onSubmit={this.onAddCategorySubmit}
                          onTextChange={this.onAddCategoryTextChange}/>
            </header>
        )
    }
}

export default Header;