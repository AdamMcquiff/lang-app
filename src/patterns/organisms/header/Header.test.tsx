import { mount, shallow } from 'enzyme';
import { when } from 'mobx';
import * as React from 'react';
import RootStore from '../../../store/RootStore';
import Header from './Header';

const store = new RootStore();

describe('Header', () => {
    const props = {
        store,
    }

    it('renders without crashing', () => {
        const component = shallow(<Header {...props} />);
        expect(component).toMatchSnapshot();
    });

    it('updates addCategoryTextValue state when TextForm value is changed', () => {
        const component = mount(<Header {...props} />);
        
        const newValue = "some value";

        component
            .find('input')
            .simulate('change', { target: { value: newValue }});

        expect(component.state().addCategoryTextValue).toEqual(newValue); 
    });

    it('adds new TranslationCategory on form submit', (done) => {
        const component = mount(<Header {...props} />);
        const instance = component.instance() as Header;
        const onAddCategorySubmitMock = jest.spyOn<any, any>(instance, 'onAddCategorySubmit');

        expect(onAddCategorySubmitMock).toHaveBeenCalledTimes(0);
        expect(component.props().store.translationCategoryStore.translationCategories.length).toEqual(0);

        component
            .find('form')
            .simulate('keydown', { key: 'Enter' });

        when(() => component.props().store.translationCategoryStore.translationCategories.length === 1, () => {
            expect(component.props().store.translationCategoryStore.translationCategories.length).toEqual(1);
            done();
        });
    });
});