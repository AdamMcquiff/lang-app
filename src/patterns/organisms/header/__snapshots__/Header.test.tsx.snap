// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders without crashing 1`] = `
<header
  className="o-header undefined"
>
  <h2>
    Categories
  </h2>
  <ul />
  <TextForm
    buttonA11yLabel="Label me."
    buttonTextLabel="Label me."
    hasButton={false}
    inputLabel="New Category"
    onSubmit={[Function]}
    onTextChange={[Function]}
  />
</header>
`;
